Service registry RHOAS CLI

PaulRMellor opened the pull request:
**Updates to the Getting Started CLI guides**

- Creates three guides as follows:
-- Installing and configuring the rhoas CLI
-- Getting started with the rhoas CLI for OpenShift Streams for Apache Kafka
-- Getting started with the rhoas CLI for OpenShift Service Registry
- Removes install steps from the _getting tarted_ guides, replaced by prereq that goes to the new common _Install_guide
- Updates links and attributes to reflect the new structure
- Removes the _Commands for..._ tables to replace with `-h` list and link to command reference guide
- Pushes the _Using_ procedures for Kafka Getting Started up a level (==) and removes the discrete title flag

wtrocki commented:
🤩

wtrocki commented:
🤩  @PaulRMellor How I can verify/render this - I'm aware of the quickstart procedure. How to run/verify tutorials?

PaulRMellor commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -0,0 +1,360 @@
+
+[NOTE]
+====
+You can use the `rhoas` CLI to manage artifacts without a service account.
+By connecting your applications to your Service Registry instance, you can retrieve artifacts and use them directly in your service.
```
Not sure about this note. Can we be clearer?

wtrocki replied:
Very good point. Makes no sense in this sentence.

```suggestion
By connecting your applications to your Service Registry instance, you can retrieve artifacts and use them directly in your service.
```

PaulRMellor commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -0,0 +1,360 @@
+  "name": "my-registry",
+  "owner": "User",
+  "status": "accepted",
+  "updated_at": "2021-09-27T15:35:01Z"
```
No region?

wtrocki replied:
No. SR is multitenant

wtrocki commented:
@PaulRMellor Amazing work! This is not only looking very good it will also:

- Remove content that was really hard for us to maintain and often get out of sync
- Provide clear path for future services.

#tech-ack

PaulRMellor commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -0,0 +1,360 @@
+[source,shell]
+----
+$ rhoas service-registry create --name my-registry
+Creating Service Registry instance with name: my-registry
```
Verify message slightly different to Kafka instance creation

wtrocki commented on consumer-configuration/README.adoc:
```diff
@@ -1,4 +1,4 @@
-////
+///
```
```suggestion
////
```

wtrocki commented:
While changes look good there are some build issues related to duplicated keys.
We need to add `-cli` suffix to them:
<img width="767" alt="Screenshot 2021-10-01 at 14 21 31" src="https://user-images.githubusercontent.com/981838/135626999-b4318a8c-9a24-4888-8436-9d33c6686f9a.png">
<img width="576" alt="Screenshot 2021-10-01 at 14 20 59" src="https://user-images.githubusercontent.com/981838/135627011-a7a98eb5-91c1-46cf-95d1-dc3901368666.png">

PaulRMellor commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -0,0 +1,371 @@
+{
+	"clientID":"srvc-acct-8c95ca5e1225-94a-41f1-ab97-aacf3df1",
+	"clientSecret":"facf3df1-ab97-2253-aa87-ab97",
+  "oauth_token_url": "https://identity.api.openshift.com/auth/realms/rhoas/protocol/openid-connect/token"
```
Just service registry with oauth? Not Kafka?

wtrocki replied:
Good catch! Not sure why we even print those outputs, but let's leave them - I tried to replicate what kafka had

wtrocki commented on rhoas-cli-kafka/README.adoc:
```diff
@@ -0,0 +1,282 @@
+[source,shell]
+----
+$ cat credentials.json
+{
+	"user":"srvc-acct-eb575691-b94a-41f1-ab97-50ade0cd1094",
+	"password":"facf3df1-3c8d-4253-aa87-8c95ca5e1225"
+}
```
```suggestion
{
	"clientID":"srvc-acct-8c95ca5e1225-94a-41f1-ab97-aacf3df1",
	"clientSecret":"facf3df1-ab97-2253-aa87-ab97",
        "oauth_token_url": "https://identity.api.openshift.com/auth/realms/rhoas/protocol/openid-connect/token"
}
```

PaulRMellor commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -0,0 +1,371 @@
+
+.Procedure
+
+* Download the artifact by specifying its ID or global ID.
```
Shall we explain why we would use one ID over the other? (Not sure)

wtrocki replied:
```suggestion
* Download the artifact by specifying Artifact ID that we specified when creating artifact
* Download the artifact by specifying Global ID retrieved from Kafka Message payload
```

PaulRMellor commented:
> While changes look good there are some build issues related to duplicated keys. We need to add `-cli` suffix to them: <img alt="Screenshot 2021-10-01 at 14 21 31" width="767" src="https://user-images.githubusercontent.com/981838/135626999-b4318a8c-9a24-4888-8436-9d33c6686f9a.png"> <img alt="Screenshot 2021-10-01 at 14 20 59" width="576" src="https://user-images.githubusercontent.com/981838/135627011-a7a98eb5-91c1-46cf-95d1-dc3901368666.png">

Doesn't the context make a difference? I've changed these two. :+1:

wtrocki commented:
> Doesn't the context make a difference? I've changed these two. 👍

Tooling validates if id's are not the same and fails. We need to try to be as specific as possible to avoid clashes.

wtrocki commented:
DO we know what this means:

```
asciidoctor: ERROR: modules/guides/proc-creating-kafka-instance-cli.adoc: line 2: level 0 sections can only be used when doctype is book
asciidoctor: ERROR: modules/guides/proc-creating-service-registry-instance-cli.adoc: line 2: level 0 sections can only be used when doctype is book
```

PaulRMellor commented:
@bibryam - Hi Bilgin. This PR has the _Getting started with the rhoas CLI for OpenShift Service Registry_ (`rhoas-cli-service-registry/readme.adoc`) and a (now) separate _Installing and configuring the rhoas CLI_ guide (`rhoas-cli/readme.adoc`)

Would appreciate a review.

bibryam commented:
@wtrocki  @PaulRMellor this look very good. Great job Wojciech!

A few notes bellow:

1)  "This example switches to the my-registry instance that was previously created." - is this needed? If there is a single instance, can that be selected by default?

2) rhoas service-registry artifact create --type=AVRO --artifact-id=my-artifact avro-userInfo.json   
Do we need to specify the type? I was expecting the type to be auto detected?

3) After a schema is updated, can we use the command for setting the compatability level and demonstrate that? 

4) @EricWittmann is RBAC feature in prod? I see role assigment demonstrated here? That will require also docs to be available IMO

wtrocki commented:
> "This example switches to the my-registry instance that was previously created." - is this needed? If there is a single instance, can that be selected by default?

I have followed up pattern we have in Kafka thinking that this was intentional to show user how to switch, but it in fact seems redundant and counterintuitive. 

> rhoas service-registry artifact create --type=AVRO --artifact-id=my-artifact avro-userInfo.json  
Do we need to specify the type? I was expecting the type to be auto detected?

Sadly we need to. Not sure why but the backend is not working well with AVRO schemas. It even has troubles with schemas that are provided in the service registry repository as examples:

https://github.com/Apicurio/apicurio-registry/tree/master/integration-tests/testsuite/src/test/resources/artifactTypes

This might be due to reliance on content type that comes from browser environment etc.? 
We did not had issues with OpenAPI but for this guide we wanted to show Avro. 

> @EricWittmann is RBAC feature in prod? I see role assigment demonstrated here? That will require also docs to be available IMO

As solution for this I think we can have this step commented out for the moment and let team to uncomment it when this is there.

PaulRMellor commented:
> > "This example switches to the my-registry instance that was previously created." - is this needed? If there is a single instance, can that be selected by default?
> 
> I have followed up pattern we have in Kafka thinking that this was intentional to show user how to switch, but it in fact seems redundant and counterintuitive. @PaulRMellor Do you think we can remove it?

>>> [PAUL] I've removed as a step, but I think it's useful to point it out. Added as a note in prereqs.

> 
> > rhoas service-registry artifact create --type=AVRO --artifact-id=my-artifact avro-userInfo.json >  > 
> > Do we need to specify the type? I was expecting the type to be auto detected?
> 
> Sadly we need to. Not sure why but the backend is not working well with AVRO schemas that are sent by CLI. It even has troubles with schemas that are provided in the service registry repository as examples:
> 
> https://github.com/Apicurio/apicurio-registry/tree/master/integration-tests/testsuite/src/test/resources/artifactTypes
> 
> This might be due to reliance on content type that comes from browser environment etc.? We did not had issues with OpenAPI but for this guide we wanted to show Avro.
> 
> > @EricWittmann is RBAC feature in prod? I see role assigment demonstrated here? That will require also docs to be available IMO
> 
> As solution for this I think we can have this step commented out for the moment and let team to uncomment it when this is there.
>>> [PAUL]  I've commented out the role creation step. We don't refer to role in the following tasks. Will we need it in the guide at all?

wtrocki commented:
> [PAUL] I've commented out the role creation step. We don't refer to role in the following tasks. Will we need it in the guide at all?

Yes.. That will be the *only* place in the docs that mention needs to apply separate role in service account. 
Apart from that we going to also add that in software:
https://github.com/redhat-developer/app-services-cli/issues/1118

wtrocki commented:
@bhardesty Good for docs review

wtrocki commented on rhoas-cli-kafka/README.adoc:
```diff
@@ -208,20 +208,19 @@ After creating a Kafka instance, you can create Kafka topics to start producing
-This example switches to the `my-kafka` instance that was previously created.
+[NOTE]
+====
+You can use `rhoas` to switch to a specific Kafka instance.
```
`rhoas` => `rhoas service-registry use`

PaulRMellor replied:
We mean the `rhoas` tool here. But your suggestion works. Updated

wtrocki commented on rhoas-cli-service-registry/README.adoc:
```diff
@@ -227,20 +230,19 @@ Artifacts might include, for example, schemas that define the structure of Kafka
-This example switches to the `my-registry` instance that was previously created.
+[NOTE]
+====
+You can use `rhoas` to switch to a specific Service Registry instance.
```
`rhoas` => `rhoas service-registry use`

bhardesty commented on rhoas-cli-kafka/README.adoc:
```diff
@@ -0,0 +1,283 @@
+
+[role="_additional-resources"]
+.Additional resources
+* {base-url-cli}{rhoas-cli-ref-url}[_CLI command reference (rhoas)_^]
```
I'm sure you're aware, but just a reminder that we'll need to define these attributes in the downstream docs repo so that the downstream version points to the CLI command ref on the customer portal.

PaulRMellor replied:
Thanks, Ben. Noted

bhardesty reviewed (approved):
@PaulRMellor these changes look great. The break down into three separate guides is a big improvement.

Just one thing to keep in mind if you haven't verified this already: once this is merged, we'll need to verify that the splitter and downstream sync job(s) handle the new guides correctly.

wtrocki commented:
I guess we good to merge?
